ca_tools/scripts/draw_square.py

```diff
@@ -0,0 +1,214 @@
+
+        self._goal_angle = self._wrapAngle(math.atan2(diff_y,diff_x))
+        self._diff_angle = self._wrapAngle(self._goal_angle - self._current_angle)
+        self._length_diff = math.sqrt(diff_x*diff_x + diff_y*diff_y)
+
+    def _setSquareGoal(self):
+            q = self._current_pose.pose.orientation
+            euler = efq((q.x,q.y,q.z,q.w))
+            euler_yaw = self._wrapAngle(euler[2] + math.pi/2.0)
+            q = qfe(euler[0],euler[1],euler_yaw) 
+            self._goal_pose.position.x = self._current_pose.pose.position.x + self._square_length*math.cos(euler_yaw)
+            self._goal_pose.position.y = self._current_pose.pose.position.y + self._square_length*math.sin(euler_yaw)
+            return 
+
+    def _getNextState(self):
+        if(self._diff_angle > ANGLE_THRESHOLD):
+            self._state.SetState("ROTATE")
+        elif(self._length_diff > LENGTH_THRESHOLD):
+            self._state.SetState("MOVE_FORWARD")
+        else:
+            self._state.SetState("STOP")
+            self._currentGoalReached = True
+
+
+    def _moveForward(self):
+        rospy.loginfo("MF")
+        t0 = rospy.get_time()
+        distance_travelled = 0
+        self._twist.linear.x = LINEAR_VEL
+        self._twist.angular.z = 0
+        while(abs(distance_travelled - self._length_diff) > LENGTH_THRESHOLD):
+            self._publish()
+            t1 = rospy.get_time()
+            distance_travelled = (self._twist.linear.x)*(t1-t0)
+        self._length_diff -= distance_travelled
+        self._stop()
+
+
+    def _updateCurrentPose(self):
+        self._current_pose = self._ground_truth.GetPose() #Could be better implemented using callbacks
+
+    def _goalReached(self):
+        rospy.loginfo("goal_reached")
+        if(self._length_diff < GOAL_DIST_THRESHOLD ):
+            self._currentGoalReached = True
+            return True
+        else:
+            return False
+
+    
+    _actions = {'MOVE_FORWARD':_moveForward ,'STOP':_stop ,'ROTATE':_rotate}
+
+    def run(self):
+        rospy.loginfo("run")
+        aux = Pose()
+        aux.position.x = -1
+        aux.position.y = -2
+        aux.position.z = 0
+        self._goal_pose = aux
+
+        while True:
+            self._updateCurrentPose()
+
+            if(self._goalReached()):
+                self._setSquareGoal()
+                self._currentGoalReached = False
+            
+            self._getPoseDiff()
+            self._getNextState()
+
+            self._actions[self._state.GetState()](self)
+
+            rospy.loginfo("First goal: %r",self._currentGoalReached)
+            rospy.loginfo("diff angle: %f",self._diff_angle)
+            rospy.loginfo("diff length :%f",self._length_diff)
+            rospy.loginfo("goal x: %f", self._goal_pose.position.x)
+            rospy.loginfo("goal y: %f", self._goal_pose.position.y)
+            for i in range(0,10):
+                self._rate.sleep()
+
+
+
+
+
+def main(stdscr):
```

Comment:
What does `stdscr` is?

Reply:
It was used with the curses module, but it was there because of an initial copy-paste from another module. In the next commit all that is related to the curses module won't be present.